pytest/conftest.py

```diff
@@ -0,0 +1,219 @@
+                "flake8": True,
+                "mypy": True,
+                "markexpr": "flake8 or mypy",
+                "reportchars": "fE",
+            }
+        )
+    if config.getoption("--demo"):
+        options.update(
+            {
+                "html": "demo.html",
+                "no_header": True,
+                "showcapture": "log",
+                "tb": "line",
+            }
+        )
+    for attr, value in options.items():
+        setattr(config.option, attr, value)
+
+
+def pytest_generate_tests(metafunc: Metafunc):
+    """Parametrize the tests based on our inputs.
+
+    Note that this hook is run for each test, so we do the file I/O in
+    'pytest_configure' and save the results.
+
+    """
+    # TODO: Provide a default target?
+    assert TARGETS, "No targets specified!"
+    if "target" in metafunc.fixturenames:
+        metafunc.parametrize("target", TARGETS, True, TARGET_IDS)
+
+
+def pytest_collection_modifyitems(
+    session: Session, config: Config, items: List[Item]
+) -> None:
+    """Pytest hook for modifying the selected items (tests).
+
+    https://docs.pytest.org/en/latest/reference.html#pytest.hookspec.pytest_collection_modifyitems
+
+    """
+    # Validate LISA mark on every item.
+    for item in items:
+        m = item.get_closest_marker("lisa")
+        assert m, f"{item} is missing required LISA marker!"
+        try:
+            lisa.validate(m)
+        except schema.SchemaMissingKeyError as e:
+            print(f"Test {item.name} failed LISA validation {e}!")
+            items[:] = []
+            return
+
+    # Optionally select tests based on a playbook.
+    included: List[Item] = []
+    excluded: List[Item] = []
+
+    # TODO: Remove logging.
+    def select(item: Item, times: int, exclude: bool) -> None:
+        """Includes or excludes the item as appropriate."""
+        if exclude:
+            print(f"    Excluding {item}")
+            excluded.append(item)
+        else:
+            print(f"    Including {item} {times} times")
+            for _ in range(times - included.count(item)):
+                included.append(item)
+
+    book = get_playbook(config.getoption("--playbook"))
+    for c in book.get("criteria"):
+        print(f"Parsing criteria {c}")
+        for item in items:
+            m = item.get_closest_marker("lisa").kwargs
+            if any(
+                [
+                    c["name"] and c["name"] in item.name,
+                    c["area"] and c["area"].casefold() == m["area"].casefold(),
+                    c["category"]
+                    and c["category"].casefold() == m["category"].casefold(),
+                    c["priority"] and c["priority"] == m["priority"],
+                    c["tags"] and set(c["tags"]) <= set(m["tags"]),
+                ]
+            ):
+                select(item, c["times"], c["exclude"])
+    if not included:
+        included = items
+    items[:] = [i for i in included if i not in excluded]
```

Comment:
Can test cases are sorted here by priority?

Reply:
Yup!

Reply:
It's just a Python list. We can sort them, delete them, add (or duplicate them) by _whatever_ criteria we could possibly want. This is where @lpereira thinks we can apply some linear discrete optimization to sort tests by cost.